crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,262 @@
+Many single-value types exist that are not singleton types: examples include
+`Literal[123456]`, `Literal[b"foo"]`, `tuple[Literal[True], Literal[False]]`, and `Literal["foo"]`.
+All runtime inhabitants of the `Literal["foo"]` type are equal to the string `"foo"`.
+However, they are not necessarily the *same* object in terms of identity;
+multiple `str` instances equal to `"foo"` can exist at runtime at different memory addresses.[^1]
+This means that it is not safe to narrow a single-value type by identity unless it is also known
+that the type is a singleton type.
+
+Single-value types in unions can be safely narrowed using inequality (but not using equality):
+
+```py
+from typing import Literal
+
+def f(x: str | Literal[1]):
+    if x != 1:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+    else:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+
+    if x == 1:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+    else:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+```
+
+An example of a singleton type that is not a single-value type
+would be `Literal[Ham.A]` for the following `Ham` enum:
+because the `Ham` class overrides `__eq__`, we can no longer say for sure
+that the `Ham` type can be safely narrowed using inequality as in the above example
+(but we *can* still narrow the `Ham` type using identity, as with all singleton types):
+
+```py
+from enum import Enum, auto
+
+class Ham(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+    def __eq__(self, other):
+        return False
+```
+
+## Final classes
+
+A "final class" in Python is a class that can be statically inferred as being unsubclassable.
+Most final classes are decorated with [`@typing.final`](https://docs.python.org/3/library/typing.html#typing.final),
+a special decorator that indicates to the type checker that the tool should assume
+the class cannot be subclassed, and that the tool should emit an error
+if it detects an attempt to do so. Not all final classes are decorated with `@final`, however.
+The most common example of this is enum classes. Any enum class that has at least one member is considered
+implicitly final, as attempting to subclass such a class will fail at runtime.
+
+For any two classes `X` and `Y`, if `X` is `@final` and `X` is not a subclass of `Y`,
+the intersection of the types `X & Y` is equivalent to `Never`, the empty set.
+This therefore means that for a type defined as being "all instances of the final class `X`",
+there are fewer ways of subtyping such a type than for most types in Python.
+However, types associated with final classes can still have non-`Never` proper subtypes.
```

Comment:
I think "types associated with final classes" is a vague and verbose way to word this; "final class types" I think is a clearer unambiguous reference to the "a type defined as ..." in the previous sentence.
```suggestion
However, final class types can still have non-`Never` proper subtypes.
```

Reply:
I guess I was going out of my way to avoid using the term "final types" here, because I was getting less and less convinced that it was actually a useful term that didn't just invite confusion 😄 But you're probably right that my roundabout language here doesn't really add any clarity.